### PR TITLE
Add navigation to search component

### LINKED ISSUE #30: Search bar with pagination
In current frontend revamp, we propose a search bar for users to search for a specific derivation in nixpkgs.

The search bar should take a derivation name as input.
Upon being filled, it queries for derivation that match it.
The list of results are displayed in a table.

(Later, the goal will be that upon clicking on a result, the graph of dependencies would be displayed on the graph view.)

Since there are many, many derivations, the search query and table of results should be paginated.

### PR DESCRIPTION
**Pull request by jlesquembre**
Closes #30

Screenshot:
localhost_5173_

The logic was getting hard to follow. To help with it, I'm using a new dependency, [xstate](https://xstate.js.org/docs/). An additional benefit is that we can visualize the state machine:

stately ai_viz_6cbc5e6c-5f62-4c89-b553-508f2f0c40c0

https://stately.ai/viz/6cbc5e6c-5f62-4c89-b553-508f2f0c40c0

### CODE REVIEW

**Comment by dorranh** [comment]
Wow, this looks great! TIL about `xstate` - it seems super useful.

File: web/src/api.ts [hunk 1]
```diff
@@ -32,7 +32,7 @@ export interface QueryResultPayload {
 export async function getPackages({
   search = "",
   limit = 10,
-  cursor = null,
+  cursor = null as Cursor,
```
Comment (GuillaumeDesforges) [inline web/src/api.ts:35]: This is quite unsettling... Why did we need this?

**Review by GuillaumeDesforges — APPROVED** [review]
Very nice!

It was a bit complex at first to get into the code as things are referenced by string, but once I got the gist it was fairly clear.

Thanks for the contribution. :+1:

**Comment by jlesquembre** [comment]
> It was a bit complex at first to get into the code as things are referenced by string, but once I got the gist it was fairly clear.

Agree, that's something to improve on xstate. You can use values instead of string, which I usually prefer, but in that case,  copy/pasting the machine schema in the visualizer doesn't always work.

File: web/src/api.ts (hunk 1 above)
Comment (jlesquembre) [reply web/src/api.ts:35]: I want to declare cursor as optional, the same with an interface:

```ts
interface IGetPackagesArgs {
  search: string;
  limit: number;
  cursor?: Cursor;
}

export async function getPackages({
  search = "",
  limit = 10,
  cursor,
}: IGetPackagesArgs): Promise<PackagesResponse> {
```

I can update it to use the interface, I like to keep the code shorter, but probably `null as Cursor` hard to read (even it took me a bit to remember what I wanted to do)

File: web/src/nix-search.ts [hunk 2]
```diff
@@ -11,50 +11,151 @@ import "@shoelace-style/shoelace/dist/components/icon/icon.js";
...
 type EventInput = Event & { target: HTMLInputElement };
 
-// TODO Add pagination
-@customElement("nix-search")
-export class NixSearch extends LitElement {
-  @state()
-  pkgs: Pkg[] = [];
-
-  @state()
+interface Context {
```
[the remark this answers is not in the document]
Comment (jlesquembre) [reply web/src/nix-search.ts:18]: We could use `type` too. Honestly, I don't know which one is better.
Looking at https://www.typescriptlang.org/play#example/types-vs-interfaces seems that it's better to use interfaces if possible.
But reading https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#differences-between-type-aliases-and-interfaces seems that you can choose based on personal preference.

[the remark this answers is not in the document]
Comment (GuillaumeDesforges) [reply web/src/nix-search.ts:18]: I think there was something about `interface` being extensible, so one can write

```typescript
interface A {
  foo: string;
}

function stuff(a: A) {
  return a.bar;
}

interface A {
  bar: string;
}
```

which I find very confusing, whereas `type` does not allow this AFAIK.

File: web/src/api.ts (hunk 1 above)
Comment (GuillaumeDesforges) [reply web/src/api.ts:35]: Got it, thanks. I think it's okay to define the type of the argument on top of the function.

The `null as Cursor` can lead to problematic "valid" type check while the program crashes runtime.

**Comment by GuillaumeDesforges** [comment]
@jlesquembre is there a blocker on the merge?
